### ISSUE TITLE
Attaching configuration to unbound ResultQuery does not work when executed async

### ISSUE THREAD

**Issue by SimonAlthoff** [jOOQ/jOOQ#10895]
### Expected behavior

During the async execution of an unbound ResultQuery via DefaultDSLContext the configuration provided by the context is used.

### Actual behavior

The previously bound configuration (in my case NoConfiguration) is used because DefaultDSLContext.fetchAsync(ResultQuery) resets the configuration before the async call is executed.

### Steps to reproduce the problem

1. Create a datasource provider
2. Create an unbound query
3. Call DSL.using(dataSourceProvider, DIALECT).fetchAsync(query)
4. Trigger actual execution

[jooq_fetch_async_test.txt](https://github.com/jOOQ/jOOQ/files/5510557/jooq_fetch_async_test.txt)

```java
package jooq.test;

import org.jooq.Record;
import org.jooq.Result;
import org.jooq.ResultQuery;
import org.jooq.SQLDialect;
import org.jooq.impl.DSL;
import org.jooq.impl.DataSourceConnectionProvider;
import org.junit.jupiter.api.Test;
import org.postgresql.ds.PGSimpleDataSource;

import javax.sql.DataSource;
import java.util.concurrent.ExecutionException;

import static org.junit.jupiter.api.Assertions.assertNotNull;

public final class JooqBindingTest {

    @Test
    public void jooqAsyncConfigurationBinding() throws ExecutionException, InterruptedException {
        final DataSource dataSource = new PGSimpleDataSource();
        final DataSourceConnectionProvider provider = new DataSourceConnectionProvider(dataSource);
        final ResultQuery<Record> query = DSL.select().from("table");
        final Result<Record> record = DSL.using(provider, SQLDialect.DEFAULT)
                .fetchAsync(query)
                .toCompletableFuture()
                .get();

        assertNotNull(record);
    }
}
```

### Versions

- jOOQ: 3.14.0
- Java: AdoptOpenJDK 11.0.8.10

**Comment by lukaseder** [comment]
Thanks a lot for your report. It makes sense, the current implementation is obviously wrong:

```java
    @Override
    public <R extends Record> CompletionStage<Result<R>> fetchAsync(ResultQuery<R> query) {
        final Configuration previous = Tools.getConfiguration(query);

        try {
            query.attach(configuration());
            return query.fetchAsync();
        }
        finally {
            query.attach(previous);
        }
    }
```

By the time you trigger the asynchronous execution, the previous configuration has been re-attached to query. This whole attaching configurations is a flawed relic from early jOOQ days, which produces mutable query instances.

Without addressing the above mutability, this issue here shouldn't be too hard to solve. We can simply delay the attaching and detaching until the actual execution.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.15.0 and 3.14.3 (#10897)
